src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,163 @@
+use crate::{email::Email, models::User, worker::Environment, Emails};
+use anyhow::anyhow;
+use crates_io_worker::BackgroundJob;
+use diesel::{
+    dsl::now, Connection, ExpressionMethods, NullableExpressionMethods, PgConnection, RunQueryDsl,
+};
+use std::sync::Arc;
+
+/// The threshold in days for the expiry notification.
+const EXPIRY_THRESHOLD: i64 = 3;
+
+/// A job responsible for monitoring the status of a token.
+/// It checks if the token is about to reach its expiry date.
+/// If the token is about to expire, the job triggers a notification.
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct CheckAboutToExpireToken;
+
+impl BackgroundJob for CheckAboutToExpireToken {
+    const JOB_NAME: &'static str = "expiry_notification";
+
+    type Context = Arc<Environment>;
+
+    #[instrument(skip(env), err)]
+    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
+        let mut conn = env.connection_pool.get()?;
+        // Check if the token is about to expire
+        // If the token is about to expire, trigger a notification.
+        check(&env.emails, &mut conn)
+    }
+}
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens =
+        crate::models::token::ApiToken::find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    // Batch send notifications in transactions.
+    const BATCH_SIZE: usize = 100;
+    for chunk in expired_tokens.chunks(BATCH_SIZE) {
+        conn.transaction(|conn| {
+            for token in chunk {
+                // Send notification.
+                let user = User::find(conn, token.user_id)?;
+                let Some(recipient) = user.email(conn)? else {
+                    return Err(anyhow!("No address found"));
+                };
```

Comment:
I think it would be more appropriate to just warn/collect the failed tokens and continue processing the remaining pending tokens, for the same reasons mentioned previously.

Reply:
though I'm not sure how failed tokens will be notified to the admin or how the admin will be able to perceive them.

Reply:
I'm not so sure about the likelihood of this happening. As I recall, we can't use the token before verifying the email. I'll check.

Reply:
I checked it. Users can generate tokens even users haven't verified their emails.

Reply:
So, in this case, is it better to log the error and continue processing the next one?

What do you think, @Turbo87 ?